src/msg_recorder/recorder.cc

```diff
@@ -10,25 +10,122 @@
+    }
+    snapshot_interval_    = recorder_config.snapshot_intervals_.back().interval_sec_;
+    snapshot_subdir_name_ = recorder_config.snapshot_intervals_.back().name_;
+    snapshot_thread_      = std::thread(std::bind(&MessageRecorder::SnapshotWorkerStart, this));
+}
+
+void MessageRecorder::SnapshotWorkerStart() {
+    while (!snapshot_shutdown_flag_.load()) {
+        auto before = std::chrono::system_clock::now();
+        MakeSnapshot();
+        std::unique_lock<std::mutex> lck(snapshot_mtx_);
+        snapshot_cv_.wait_until(lck, before + snapshot_interval_, [this] { return snapshot_shutdown_flag_.load(); });
+    }
+}
+
+void MessageRecorder::MakeSnapshot() {
+    std::unique_lock<std::mutex> lock(snapshot_mtx_);
+    snapshot_pause_flag_.store(false);
+    AddJobToRunner(
+        [this]() {
+            GenerateSnapshot(keep_max_);
+            snapshot_pause_flag_.store(true);
+            snapshot_cv_.notify_all();
+        },
+        record_strand_);
+
+    // wait until the current generation to be finished
+    snapshot_cv_.wait(lock, [this] { return snapshot_pause_flag_.load(); });
+}
+
+void MessageRecorder::SnapshotWorkerEnd() {
+    {
+        std::unique_lock<std::mutex> lock(snapshot_mtx_);
+        snapshot_shutdown_flag_.store(true);
+    }
+    snapshot_cv_.notify_all();
+    if (snapshot_thread_.joinable()) {
+        snapshot_thread_.join();
+    }
+    snapshots_.clear();
 }
 
 MessageRecorder::~MessageRecorder() {
+    SnapshotWorkerEnd();
     files_.clear();
-
     if (std::filesystem::is_empty(GetRecordDir())) {
         LOG(INFO) << "Record dir " << GetRecordDir() << " is empty, removing...";
         std::filesystem::remove(GetRecordDir());
     }
 }
 
+void MessageRecorder::StopMainLoop() {
+    SnapshotWorkerEnd();
+}
+
+void MessageRecorder::GenerateSnapshot(const std::size_t max) {
+    for (const auto& file : files_) {
+        file->CloseDB();
+    }
+
+    // create dir for individual interval
+    std::filesystem::path interval_dir = record_dir_.parent_path() / snapshot_dir_name_ / snapshot_subdir_name_;
+    std::filesystem::create_directories(interval_dir);
+
+    const auto snapshot_dir = interval_dir / SnapshotDirNameGenerator();
+    const auto options      = std::filesystem::copy_options::recursive | std::filesystem::copy_options::skip_symlinks;
+    std::filesystem::create_directories(snapshot_dir);
+    std::filesystem::copy(record_dir_, snapshot_dir, options);
+
+    snapshots_.push_back(snapshot_dir);
+
+    while (snapshots_.size() > max) {
+        auto old_snapshot = snapshots_.front();
+        if (std::filesystem::exists(old_snapshot)) {
+            std::uintmax_t n = std::filesystem::remove_all(old_snapshot);
+            if (n > 0) {
+                LOG(INFO) << "Removed " << n << " files or directories";
+            } else {
+                LOG(ERROR) << "Latest snapshot path fail to remove, something's wrong";
```

Comment:
- What is wrong?
- "Something's" -> "something is". Please write formally in logs.